authentication/account/service/user.go

```diff
@@ -112,10 +115,43 @@ func (s *userServiceImpl) BanUser(ctx context.Context, username string) (*reposi
 	return identity, err
 }
 
-func (s *userServiceImpl) ListIdentitiesToNotifyBeforeDeactivation(ctx context.Context) ([]repository.Identity, error) {
-	since := time.Now().Add(time.Duration(-s.config.GetUserDeactivationInactivityNotificationPeriod()*24) * time.Hour) // remove 'n' days from now
+// NotifyIdentitiesBeforeDeactivation list identities (with a limit) who are eligible for account deactivation,
+// sends a notification to each one and record the timestamp of the notification as a marker before upcoming deactivation
+func (s *userServiceImpl) NotifyIdentitiesBeforeDeactivation(ctx context.Context) ([]repository.Identity, error) {
+	since := time.Now().Add(time.Duration(-s.config.GetUserDeactivationInactivityNotificationPeriod()) * 24 * time.Hour) // remove 'n' days from now
 	limit := s.config.GetUserDeactivationFetchLimit()
-	return s.Repositories().Identities().ListIdentitiesToNotifyForDeactivation(ctx, since, limit)
+	identities, err := s.Repositories().Identities().ListIdentitiesToNotifyForDeactivation(ctx, since, limit)
+	if err != nil {
+		return nil, errs.Wrap(err, "unable to send notification to users before account deactivation")
+	}
+	// for each identity, send a notification and record the timestamp in a separate transaction.
+	// perform the task for each identity in a separate Tx, and just log the error if something wring happen,
+	// but don't stop processing on the rest of the accounts.
+	expirationDate := time.Now().Add(time.Duration(s.config.GetUserDeactivationInactivityPeriod()-s.config.GetUserDeactivationInactivityNotificationPeriod()) * 24 * time.Hour).Format("Mon Jan 2")
+	notificationDate := time.Now()
+	for _, identity := range identities {
+		msg := notification.NewUserDeactivationEmail(identity.ID.String(), identity.Username, expirationDate)
+		_, err := s.Services().NotificationService().SendMessageAsync(ctx, msg)
```

Comment:
We should be careful here. We can easily end up with sending a thousand (and even more in the first run) notifications every day. Would need to make sure that there is no problems with creating thousands of go routine to send notification requests at once. Maybe we should add some grace before sending another notification. Something like 0.5 sec?

cc: @sbryzak @MatousJobanek

Reply:
good point. Let me add a configurable delay here.

Reply:
@alexeykazakov addressed in b6a24ed